src/sbml/conversion/SBMLRateRuleConverter.cpp

```diff
@@ -0,0 +1,1090 @@
+    if (node->getNumChildren() != 2)
+        return false;
+
+    // if node is not a minus, it's not "-" in k-x
+    if (node->getType() != ASTNodeType_t::AST_MINUS)
+        return false;
+
+    // if right child is not a variable species, it's not k-x
+    ASTNode* rightChild = node->getRightChild();
+    if (!isVariableSpecies(rightChild, model))
+        return false;
+
+    // if left child is not a numerical constant or a parameter, it's not k-x - else it is!
+    ASTNode* leftChild = node->getLeftChild();
+    return isNumericalConstantOrParameter(leftChild, model);
+}
+
+bool SBMLRateRuleConverter::isVariableSpecies(ASTNode* node, Model* model)
+{
+    if (!node->isName()) // some nodes, like * operators, don't seem to have a name in the first place
+        return false;
+    Species* species = model->getSpecies(node->getName());
+    bool isSpecies = (species != NULL);
+    bool isVariable = !node->isConstant();
+    return isVariable && isSpecies;
+}
+
+bool SBMLRateRuleConverter::isNumericalConstantOrParameter(ASTNode* node, Model* model)
+{
+    if (!node->isName()) // some nodes, like * operators, don't seem to have a name in the first place
+        return false;
+    Parameter* parameter = model->getParameter(node->getName());
+    bool isParameter = (parameter != NULL);
+    bool isNumericalConstant = node->isNumber() && node->isConstant();
+    return isNumericalConstant || isParameter;
+}
+
+void
+SBMLRateRuleConverter::dealWithSpecies()
+{
+  // loop through ODE pairs and check that any variables are listed as Species
+  for (odeIt it = mODEs.begin(); it != mODEs.end(); ++it)
+  {
+    const std::string variable = it->first;
+    // already a species
+    if (mDocument->getModel()->getSpecies(variable) != NULL)
+    {
+      continue;
+    }
+    else
+    {
+      // must be a parameter
+      Parameter *p = mDocument->getModel()->removeParameter(variable);
+      Species *newSpecies = mDocument->getModel()->createSpecies();
+      newSpecies->setId(variable);
+      newSpecies->setInitialAmount(p->getValue());
+      newSpecies->setHasOnlySubstanceUnits(true);
+      newSpecies->setBoundaryCondition(false);
+      newSpecies->setConstant(false);
+      if (mDocument->getModel()->getCompartment(0) == NULL)
+      {
+        Compartment * newCompartment = mDocument->getModel()->createCompartment();
+        newCompartment->setId("addedCompartment");
+        newCompartment->setConstant(true);
+        newCompartment->setSize(1.0);
+        newSpecies->setCompartment("addedCompartment");
+      }
+      else
+      {
+        newSpecies->setCompartment(mDocument->getModel()->getCompartment(0)->getId());
+      }
+      delete p;
+    }
+  }
+}
+void
+SBMLRateRuleConverter::createReactions()
+{
+  unsigned int i = 0;
+  for (setCoeffIt it = mCoefficients.begin(); it != mCoefficients.end(); ++it)
+  {
+    Reaction *r = mDocument->getModel()->createReaction();
+    r->setReversible(false);
```

Comment:
please give the newly created reaction an id. any id will do. Leaving it empty will cause invalid models to be created. Some basic heuristic would be to use `J` + number of reactions in the model, or having that counter jut increase in case such an entity already exists in the model.